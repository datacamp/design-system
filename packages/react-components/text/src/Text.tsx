import './sideEffects/injectRegularFontFace';

import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css } from '@emotion/core';
import { mapKeys } from 'lodash';
import React from 'react';

interface TextProps {
  children: string;
  className?: string;
  dataAttributes?: { [key: string]: string };
}

const baseTextStyle = css({
  color: tokens.color.opaque.grey.value.rgb,
  fontFamily: [
    tokens.asset.font.sansSerif.attributes.fallback,
    tokens.asset.font.sansSerif.value,
  ],
  fontSize: tokens.size.font.base.value,
  fontStyle: 'normal',
  fontWeight: tokens.size.fontWeight.regular.value,
  lineHeight: tokens.size.font.base.value,
});

export const Text: React.FC<TextProps> = ({
  children,
  className,
  dataAttributes,
}: TextProps) => {
  const parsedDataAttributes = mapKeys(
    dataAttributes,
    (_, key) => `data-${key}`
  );

  return (
    <span css={baseTextStyle} className={className} {...parsedDataAttributes}>
      {children}
    </span>
  );
};

export default Text;
