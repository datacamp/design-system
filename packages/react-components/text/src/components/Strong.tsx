import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css } from '@emotion/core';
import React from 'react';

import baseStyle from '../baseStyle';
import computeDataAttributes from '../computeDataAttributes';

interface StrongProps {
  children: string;
  className?: string;
  dataAttributes?: { [key: string]: string };
}

const strongStyle = css(baseStyle, {
  fontWeight: tokens.size.fontWeight.bold.value,
  lineHeight: tokens.size.font.base.value,
});

const Strong: React.FC<StrongProps> = ({
  children,
  className,
  dataAttributes,
}) => {
  const parsedDataAttributes = computeDataAttributes(dataAttributes);

  return (
    <strong className={className} css={strongStyle} {...parsedDataAttributes}>
      {children}
    </strong>
  );
};

export default Strong;
