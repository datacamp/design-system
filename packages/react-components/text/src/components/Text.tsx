import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css } from '@emotion/core';
import React from 'react';

import baseStyle from '../baseStyle';
import computeDataAttributes from '../computeDataAttributes';
import validateChildrenProp from '../validateChildrenProp';
import Small from './Small';
import Strong from './Strong';

export interface TextProps {
  children: string | (string | React.ReactElement)[];
  className?: string;
  dataAttributes?: { [key: string]: string };
}

const textStyle = css(baseStyle, {
  fontSize: tokens.size.font[300].value,
  fontWeight: tokens.size.fontWeight.regular.value,
  lineHeight: tokens.size.font.base.value,
});

const Text: React.FC<TextProps> = (props: TextProps) => {
  validateChildrenProp(props, 'Text', [Small, Strong]);
  const { children, className, dataAttributes } = props;
  const parsedDataAttributes = computeDataAttributes(dataAttributes);

  return (
    <span className={className} css={textStyle} {...parsedDataAttributes}>
      {children}
    </span>
  );
};

export default Text;
