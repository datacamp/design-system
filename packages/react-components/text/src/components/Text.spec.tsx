import 'jest-dom/extend-expect';

import { render } from '@testing-library/react';
import React from 'react';

import Strong from './Strong';
import Text from './Text';

const testText = 'Some example text.';
const testClassName = 'example-class';
const testDataAttributes = { cy: 'othertest', example: 'testdata' };

describe('<Text />', () => {
  it('renders with a string', () => {
    const { container } = render(
      <Text className={testClassName} dataAttributes={testDataAttributes}>
        {testText}
      </Text>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with Strong children', () => {
    const { container } = render(
      <Text className={testClassName} dataAttributes={testDataAttributes}>
        Some example text <Strong>with strong</Strong> included.
      </Text>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('sets the text in the span', () => {
    const { getByText } = render(<Text>{testText}</Text>);
    expect(getByText(testText)).toBeDefined();
  });

  it('sets the className', () => {
    const { getByText } = render(
      <Text className={testClassName}>{testText}</Text>
    );
    expect(getByText(testText)).toHaveClass(testClassName);
  });

  it('sets the data attributes', () => {
    const { getByText } = render(
      <Text dataAttributes={testDataAttributes}>{testText}</Text>
    );
    expect(getByText(testText)).toHaveAttribute(
      'data-cy',
      testDataAttributes.cy
    );
    expect(getByText(testText)).toHaveAttribute(
      'data-example',
      testDataAttributes.example
    );
  });
});
