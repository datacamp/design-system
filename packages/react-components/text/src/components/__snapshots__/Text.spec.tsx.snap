// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> renders with Small children 1`] = `
.emotion-1 {
  -webkit-font-smoothing: antialiased;
  color: rgb(5,25,45);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
}

.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(5,25,45);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-size: 14px;
  line-height: 16px;
}

<span
  class="example-class emotion-1"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text 
  <small
    class="emotion-0"
  >
    with strong
  </small>
   included.
</span>
`;

exports[`<Text /> renders with Strong children 1`] = `
.emotion-1 {
  -webkit-font-smoothing: antialiased;
  color: rgb(5,25,45);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
}

.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(5,25,45);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-weight: 800;
  line-height: 16px;
}

<span
  class="example-class emotion-1"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text 
  <strong
    class="emotion-0"
  >
    with strong
  </strong>
   included.
</span>
`;

exports[`<Text /> renders with a string 1`] = `
.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(5,25,45);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
}

<span
  class="example-class emotion-0"
  data-cy="othertest"
  data-example="testdata"
>
  Some example text.
</span>
`;
